src/ORM/DB.php

```diff
@@ -109,6 +126,124 @@ public static function get_conn($name = 'default')
         return null;
     }
 
+    /**
+     * Only use the primary database connection for the rest of the current request
+     *
+     * @internal
+     */
+    private static bool $mustUsePrimary = false;
+
+    /**
+     * Used by DB::withPrimary() to count the number of times it has been called
+     * Uses an int instead of a bool to allow for nested calls     *
+     *
+     * @internal
+     */
+    private static int $withPrimaryCount = 0;
+
+    /**
+     * The key of the previously choosen random replica config
+     *
+     * @internal
+     */
+    private static string $randomReplicaConfigKey = '';
+
+    /**
+     * Flag to determine if replicas can be used
+     *
+     * @internal
+     */
+    private static bool $canUseReplicas = true;
+
+    /**
+     * Set the flag to allow or disallow the use of replica database connections
+     * For instance disable when running unit-tests
+     */
+    public static function setCanUseReplicas(bool $canUseReplicas): void
+    {
+        DB::$canUseReplicas = $canUseReplicas;
+    }
+
+    /**
+     * Set the flag to only use the primary database connection for the current request
+     * meaning that replia connections will no longer be used
+     *
+     * This intentioally does not have a parameter to set this back to false, as this it to prevent
+     * accidentally attempting writing to a replica, or reading from an out of date replica
+     * after a write
+     */
+    public static function setMustUsePrimary(): void
+    {
+        DB::$mustUsePrimary = true;
+    }
```

Comment:
I've intentionally not set a getter here because I think it's a foot gun. There's so many weird edge cases that can happen with DB replicas being out of sync I think we should be very conservative here and simply force the use of the primary once a single write operation has occurred and not allow people to change that behavior.

> use cached data for heavy queries when using the primary, but fetch it fresh when using replicas

This idea sounds entirely backwards? cached data for heavy queries? The idea with replicas is you do you use the replicas for performant read operations and the primary for writes and times where you cannot tolerate out of sync data

Reply:
A getter just tells you what the value is. It doesn't change the value. I don't see how telling people what the current value is would cause problems like what you're describing?